docs/design/variadics.md

```diff
@@ -0,0 +1,1018 @@
+-   _Parameterized type equality:_ Let `P` be a parameterized type. `P(A, B)`
+    equals `P(C, D)` if `A` equals `C` and `B` equals `D`.
+
+This is complicated by the fact that in a function call, the parameter tuple
+type may use deduced parameters of the function, which act as unknowns that we
+must solve for. We will model parameter deduction using the concept of a
+_binding map_, which is a set of pairs where the first element of the pair is a
+deduced parameter, and the second element is the deduced value of that
+parameter. A binding map must be _consistent_, which means that it cannot
+contain two pairs with the same first element (for simplicity, we will require
+function signatures and other full patterns to be expressed without deduced
+parameters that have pack names when applying these rules, so we don't need to
+worry about pairs with overlapping but non-equal first elements).
+
+We can then re-express convertibility and equality as ternary relations: "`T` =
+`U` given a binding map `M`" means that if we rewrite `T` and `U` by replacing
+every deduced parameter with the corresponding value in `M`, the two rewritten
+forms are equal. "`T` is convertible to `U` given a binding map `M`" is
+interpreted similarly. For example, we can re-express the last deduction rule
+above in terms of these ternary relations as follows:
+
+-   _Parameterized type equality:_ Let `P` be a parameterized type. If `A`
+    equals `C` given a binding map `M1`, and `B` equals `D` given a binding map
+    `M2`, and `M1` ∪ `M2` is consistent, then `P(A, B)` equals `P(C, D)` given a
+    binding map `M1` ∪ `M2`.
+
+Almost all of the deduction rules can be generalized in the same way: the
+binding map in the conclusion is the union of the binding maps in the premises,
+and there's an additional premise that this union is consistent. If there are no
+"equals" or "convertible to" relations in the premises, the binding map in the
+conclusion is empty. When that pattern applies, we will usually leave the
+binding maps implicit (as in the original formulation of the above rule). One
+major exception is the following rule:
+
+_Binding introduction:_ Let `X` be a deduced parameter with type `T`. If the
+type of `E` is convertible to `T` given binding map `M`, and `M` ∪ {(`X`, `E`)}
+is consistent, then `X` equals `E` given a binding map `M` ∪ {(`X`, `E`)}.
+
+The deduction rules that are specific to variadic types are as follows:
+
+-   _Expansion convertibility:_ `...T` is convertible to `...U` if the arity of
+    `U` equals the arity of `T`, and all scalar components of `T` are
+    convertible to all scalar components of `U`.
+-   _Shape equality:_ Let `(S1s)`, `(S2s)`, `(S3s)`, and `(S4s)` be shapes.
+    `(S1s, S2s)` equals `(S3s, S4s)` if `(S1s)` equals `(S3s)` and `(S2s)`
+    equals `(S4s)`.
+
+#### Example
+
+Suppose we are typechecking the following function call:
+
+```carbon
+fn F[T:! type](x: Vector(T));
+F(() as Vector(i32));
+```
```

Comment:
I think there has been some scope creep here: you are defining deduction in general, independent of variadics.

Reply:
OK, I've removed the non-variadic-specific parts.

(To be clear, "scope creep" wasn't how this happened; earlier versions of the design made more invasive changes to the convertibility and deduction rules, so they couldn't be explained as a couple of additional rules on top of an unstated baseline)